### PR TITLE
Checked edge cases and added more tests for validating user input

### PR DESCRIPTION
**Pull request by mrfishball**

### CODE REVIEW

File: lib/board.ex [hunk 1]
```diff
@@ -1,7 +1,5 @@
 defmodule Board do
-  @moduledoc """
-    This is the game's visual engine.
-  """
+  @moduledoc false
```
Comment (cmvandrevala) [inline lib/board.ex]: Do you need `@moduledoc false` here? Can't you just set a flag on credo telling it to ignore this rule?

File: lib/board.ex [hunk 2]
```diff
@@ -1,7 +1,5 @@
...
   def new do
     for col <- 0..2, row <- 0..2, into: %{}, do: {{col, row}, " "}
```
Comment (cmvandrevala) [inline lib/board.ex]: You should refactor this method to get rid of the hard-coded values of `0..2`

File: lib/ttt.ex [hunk 3]
```diff
@@ -53,7 +51,7 @@ defmodule TTT do
   end
 
   def get_input(turn) do
-    move = IO.gets "'#{turn}', please enter a number from 1 to 9 only:"
+    move = IO.gets "'#{turn}', please enter a number from 1 to 9 only: "
     case not valid_input?(move) do
```
Comment (cmvandrevala) [inline lib/ttt.ex]: Why are you using `case not valid_input?` rather than just `case valid_input`? That makes the code harder to read.

File: test/ttt_test.exs [hunk 4]
```diff
@@ -18,20 +18,35 @@ defmodule TTTTest do
     input4 = ",,"
     input5 = "10"
     input6 = "0"
-    input7 = "5"
+    input7 = "9"
```
Comment (cmvandrevala) [inline test/ttt_test.exs]: Each of these inputs test different things. You should refactor each of these into a separate test. You might have 14 different tests since you are testing 14 different validation cases.

File: test/ttt_test.exs [hunk 5]
```diff
@@ -18,20 +18,35 @@ defmodule TTTTest do
...
   test "Valid user input will be map to the correct coordinate on the board" do
-    TTT.match_input("1") == {0, 0}
-    TTT.match_input("9") == {2, 2}
-    TTT.match_input("5") == {1, 1}
+    assert TTT.match_input("1") == {0, 0}
+    assert TTT.match_input("9") == {2, 2}
+    assert TTT.match_input("5") == {1, 1}
   end
 
   test "Visual board is updated accordingly" do
```
Comment (cmvandrevala) [inline test/ttt_test.exs]: This test name "Visual board is updated accordingly" is very unclear. Consider renaming the test to describe exactly what is being tested.

**Review by cmvandrevala — CHANGES_REQUESTED** [review]

**Comment by cmvandrevala** [comment]
One of the methods in your project is:

```
  defp put(board, {col, row}, val) do
    Map.put(board, {col, row}, val)
  end
```

Creating wrappers around core methods is a good idea, but I notice that you are literally repeating the method signature here (same arguments, same expected output). What is the point of this method? Why even have `put` if it is literally just `Map.put`?

I think wrapper methods like this are useful, but only when they abstract away implementation details. For example:

```
  defp putXValue(board, {col, row}) do
    Map.put(board, {col, row}, :x)
  end
```

I would not recommend implementing this specifically, but notice how putXValue is abstracting away the logic needed to add an x value on the board.

**Comment by cmvandrevala** [comment]
Your test suite contains a lot of tests with multiple set-ups and multiple assertions. For example:

```
  test "player wins horizontally" do
    game = Game.new
    game2 = Game.new
    game3 = Game.new

    {:ok, game} = Game.play_turn(game, :o, {0, 0})
    {:ok, game} = Game.play_turn(game, :x, {2, 1})
    {:ok, game} = Game.play_turn(game, :o, {1, 0})
    {:ok, game} = Game.play_turn(game, :x, {1, 1})
    {:ok, game} = Game.play_turn(game, :o, {2, 0})

    assert Game.status(game) == {:ended, {:winner, :o}}

    {:ok, game2} = Game.play_turn(game2, :o, {0, 1})
    {:ok, game2} = Game.play_turn(game2, :x, {2, 2})
    {:ok, game2} = Game.play_turn(game2, :o, {1, 1})
    {:ok, game2} = Game.play_turn(game2, :x, {1, 0})
    {:ok, game2} = Game.play_turn(game2, :o, {2, 1})

    assert Game.status(game2) == {:ended, {:winner, :o}}

    {:ok, game3} = Game.play_turn(game3, :o, {0, 2})
    {:ok, game3} = Game.play_turn(game3, :x, {2, 0})
    {:ok, game3} = Game.play_turn(game3, :o, {1, 2})
    {:ok, game3} = Game.play_turn(game3, :x, {1, 0})
    {:ok, game3} = Game.play_turn(game3, :o, {2, 2})

    assert Game.status(game3) == {:ended, {:winner, :o}}
  end
```

These should really be three different tests with three different test descriptions.

File: test/ttt_test.exs [hunk 6]
```diff
@@ -18,20 +18,35 @@ defmodule TTTTest do
...
+    input7 = "9"
+    input8 = "@@@"
+    input9 = "哈"
+    input10 = "-1"
+    input11 = "   1   "
+    input12 = "\#{1+1}"
+    input13 = "6666666666666666666666666"
+    input14 = "\n"
+
     assert TTT.valid_input?(input1) == :true
```
Comment (cmvandrevala) [inline test/ttt_test.exs]: Why is `TTT.valid_input?` returning `:true` and `:false`? Why not literally just return the booleans true and false instead of atoms?

**Review by cmvandrevala — CHANGES_REQUESTED** [review]

**Comment by cmvandrevala** [comment]
Add `.DS_Store` to your `.gitignore`

File: test/ttt_test.exs [hunk 7]
```diff
@@ -3,38 +3,56 @@ defmodule TTTTest do
...
+    inputs = ["1", "2", "5", "8", "9"]
+    for input <- inputs, do: assert TTT.valid_input?(input) == true
+  end
+
+  test "Whitespaces or newlines before or/and after a valid input will be ignored, such input will be a valid one" do
+    inputs = ["  3  ", "    7", "4 ", "\n2\n\n"]
+    for input <- inputs, do: assert TTT.valid_input?(input) == true
+  end
+
+  test "A number smaller than '1' or greater than '9' will be invalidated" do
```
Comment (cmvandrevala) [inline test/ttt_test.exs]: Imagine that this test were to fail. You are currently testing two assertions - numbers greater than nine and less than one. Thus, you would have no idea whether the number less than one was making the test fail or the number greater than nine was making it fail. Each logical assertion should be its own test.

File: test/ttt_test.exs [hunk 8]
```diff
@@ -3,38 +3,56 @@ defmodule TTTTest do
...
-    assert TTT.valid_input?(input4) == :false
-    assert TTT.valid_input?(input5) == :false
-    assert TTT.valid_input?(input6) == :false
-    assert TTT.valid_input?(input7) == :true
+  test "A number '1' through '9' will be a valid input" do
+    inputs = ["1", "2", "5", "8", "9"]
+    for input <- inputs, do: assert TTT.valid_input?(input) == true
+  end
+
+  test "Whitespaces or newlines before or/and after a valid input will be ignored, such input will be a valid one" do
```
Comment (cmvandrevala) [inline test/ttt_test.exs]: This should be split into at least two or three tests - whitespace before, whitespace after, newlines after, etc

File: test/ttt_test.exs [hunk 9]
```diff
@@ -3,38 +3,56 @@ defmodule TTTTest do
...
-    TTT.match_input("1") == {0, 0}
-    TTT.match_input("9") == {2, 2}
-    TTT.match_input("5") == {1, 1}
+    assert TTT.match_input("1") == {0, 0}
+    assert TTT.match_input("9") == {2, 2}
+    assert TTT.match_input("5") == {1, 1}
   end
 
-  test "Visual board is updated accordingly" do
+  test "Board is updated and rendered correctly" do
```
Comment (cmvandrevala) [inline test/ttt_test.exs]: "Board is updated and rendered correctly" - This test name does not give me much in the way of details. Under what conditions is this true? Does it ever throw an exception or fail? A better name might be:

"Board is updated and rendered correctly when user provides valid input"

Or something like that

File: test/board_test.exs [hunk 10]
```diff
@@ -22,7 +22,7 @@ defmodule BoardTest do
     board = Board.new
     coords = [{0, 2}, {1, 0}, {2, 1}, {1, 2}]
     token = :x
-    assert Board.fill(board, coords, token) == %{
+    assert Board.update(board, coords, token) == %{
```
Comment (cmvandrevala) [inline test/board_test.exs]: The name `update` is very generic. How about `Board.take_move`? Or something similar?

File: lib/board.ex [hunk 11]
```diff
@@ -28,10 +22,10 @@ defmodule Board do
     IO.puts "\n"
   end
 
-  def update(board, list, list2, token, token2) do
+  def render(board, list, list2, token, token2) do
```
Comment (cmvandrevala) [inline lib/board.ex]: These argument names are not great. I get `board`, and `token`/`token2` is ok, but what is `list`/`list2`?

File: test/board_test.exs [hunk 12]
```diff
@@ -48,7 +48,7 @@ defmodule BoardTest do
     token = :o
     opponent_token = :x
     assert capture_io(fn ->
-      Board.update(board, move, opponent_moves, token, opponent_token) end) ==
+      Board.render(board, move, opponent_moves, token, opponent_token) end) ==
```
Comment (cmvandrevala) [inline test/board_test.exs]: Why does `Board.render` require five arguments? Not saying you have to address this particular comment, but it does seem kind of weird to me...

File: test/game_test.exs [hunk 13]
```diff
@@ -36,10 +36,8 @@ defmodule GameTest do
     assert Game.play_turn(game, :x, {2, 2}) == {:error, :not_your_turn}
   end
 
-  test "player wins horizontally" do
+  test "player wins horizontally (top line)" do
```
Comment (cmvandrevala) [inline test/game_test.exs:39]: Much better!

**Review by cmvandrevala — CHANGES_REQUESTED** [review]

**Review by cmvandrevala — APPROVED** [review]
